lib/authstore/index.js

```diff
@@ -0,0 +1,1263 @@
+      deviceIndex,
+      links,
+    })
+  }
+
+  /**
+   * @param {Object} options
+   * @param {String} options.identityId
+   * @returns {Promise<DeviceStatement>}
+   */
+  async restoreDevice(options) {
+    const { identityId } = options
+
+    // order blocks by deviceIndex so we can check that they were created in the correct order
+    const existing = (
+      await this.getBlocksByType({ dataType: 'devices', identityId })
+    ).sort((a, b) => {
+      return b.deviceIndex - a.deviceIndex
+    })
+
+    // TODO: more checks to ensure timestamp/deviceIndex/authorIndex/etc. is correct
+    const [mostRecent] = existing
+
+    if (mostRecent.action !== 'device:remove') {
+      throw new Error(
+        'Device cannot be restored because it has not been removed'
+      )
+    }
+
+    if (!existing.length) {
+      throw new Error(
+        'Device cannot be restored because it has not been added or removed'
+      )
+    }
+
+    const authorIndex = await this.getAuthorIndex(this.identityId)
+    const deviceIndex = await this.getDeviceIndex(identityId)
+
+    const authorRole = await this.getRole({ identityId: this.identityId })
+
+    if (!authorRole) {
+      throw new Error('Author does not have a role')
+    }
+
+    await this.verifyRole(authorRole)
+
+    const role = this.#availableRoles.find(
+      (role) => role.name === authorRole.role
+    )
+
+    if (!role) {
+      throw new Error('Author role not found')
+    }
+
+    if (!role.capabilities.includes('manage:devices')) {
+      throw new Error('Author does not have permission to add device')
+    }
+
+    const link = await this.getBlockByDeviceIndex({
+      deviceIndex: deviceIndex - 1,
+      identityId,
+    })
+
+    if (!link) {
+      throw new Error('Previous statement for this device not found')
+    }
+
+    const links = [link.version]
+
+    await this.verifyLinks(links)
+
+    const timestamp = new Date().getTime()
+    const signature = this.#signDeviceMessage({
+      identityId,
+      authorId: authorRole.id,
+      action: 'device:restore',
+      authorIndex,
+      deviceIndex,
+      timestamp,
+      links,
+    })
+
+    return this.devices.create({
+      id: identityId,
+      type: 'devices',
+      action: 'device:restore',
+      created: timestamp,
+      signature,
+      authorIndex,
+      deviceIndex,
+      links,
+    })
+  }
+
+  /**
+   * @param {Object} options
+   * @param {String} options.role
+   * @param {String} options.identityId
+   * @returns {Promise<RoleStatement>}
+   * @throws {Error}
+   */
+  async setRole(options) {
+    const { role, identityId } = options
+
+    const existing = await this.getRole({ identityId })
+    const authorRole = await this.getRole({ identityId: this.identityId })
+
+    if (existing && existing.role === role) {
+      throw new Error(`Role ${role} already set`)
+    }
+
+    if (existing && existing.role === 'project-creator') {
+      // TODO: we may want to allow this in the future
```

Comment:
should probably remove this todo? for now things rely on a single project creator and removing them or changing their role means forking to a new project that retains existing data.